docs/proposal/aws-resource-handling.md

```diff
@@ -0,0 +1,203 @@
+  - client-go
+    - https://github.com/kubernetes/client-go/blob/master/rest/request.go#L659
+- Identify which resources fall into which class of workflow based on tagging support, client token support, etc.
+- Better define mutatable/non-mutatable attributes for objects during update
+
+## Using client tokens
+
+Where possible use [client tokens](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/Run_Instance_Idempotency.html) in the create request so that subsequent requests will return the same response.
+
+## Tagging of resources
+
+Resources that are managed by the controllers/actuators should be tagged with: `kubernetes.io/cluster/<name or id>=owned` and `sigs.k8s.io/cluster-api-provider-aws=true`. The latter tag being used to differentiate from resources managed by other tools/components that make use of the common tag.
+
+TODO: Define additional tags that can be used to provide additional metadata about the resource configuration/usage by the actuator. This is would allow us to rebuild status without relying on polluting the object config.
+
+## Handling of AWS api errors
+
+Each resource has specific error codes that it will return and these can be used to differentiate fatal errors from retryable errors. These errors are well documented in some cases (elbv2 api), and poorly in others (ec2 api). We should provide a best effort to [properly handle these errors](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/handling-errors.html) in the correct manner.
+
+## Proposed workflows
+
+### Resources that support tag on create support (with or without client tokens)
+
+#### Create
+
+- Query resource by tags to determine if resource already exists
+- Create the resource if it doesn't already exist
+- Update the cluster/machine object config and status
+  - If update fails return a retryable error to requeue the create
+- Enqueue cluster/machine update if not already available/ready
+
+TODO: flowchart
+
+#### Update
+
+- Query resource by ID
+- Update object status
+- Enqueue cluster/machine update if not already available/ready
+
+#### Edge case coverage
+
+1. Yes - tagging is handled on creation
+2. Yes - since resources are tagged on creation, returning an error and requeueing the create will find the tagged resource and attempt to retry the object update.
+3. Yes - there is no delete attempt since we can re-query the resource by tags.
+4. Yes - the next attempt to create the resource will find the already created resource by tags.
+
+### Resources that support client tokens but require separate tagging
+
+#### Create
+
+- Create the resource using object uid as the client token
+- Update the cluster/machine object config and status
+  - If update fails return a retryable error to requeue the create
+- Tag AWS resource
+- Enqueue cluster/machine update if not already available/ready
+
+TODO: flowchart
+
+#### Update
+
+- Query resource by ID
+- tag resource if missing tags
```

Comment:
This might cause problems for `shared` resources like a VPC used for multiple clusters. kubernetes core is using `kubernetes.io/cluster/<name or id>=owned` and `kubernetes.io/cluster/<name or id>=shared` to differentiate and store whether it created the resource itself or uses a shared one.

Problems:
1. Resources can only have up to 50 tags, so the re-usability of a VPC might be limited if every cluster adds a `shared` tag
1. If a resource is encountered without tag is discovered, should cluster-api assume ownership or fail? [kops](https://github.com/kubernetes/kops/blob/af4df08b694e2a1f8814a7b3649060477be67c86/docs/run_in_existing_vpc.md#running-in-a-shared-vpc) and [kubespray](https://github.com/kubernetes-incubator/kubespray/blob/e3d562bcdb0da3bc31c079907c9e04e0c1a43c68/docs/aws.md#aws) require explicit tagging of shared resources, which solves the problem but might be bad user experience.